contracts/token/ERC20/SnapshotToken.sol

```diff
@@ -0,0 +1,70 @@
+pragma solidity ^0.4.23;
+
+import { StandardToken } from "./StandardToken.sol";
+import { ArrayUtils } from "../../utils/ArrayUtils.sol";
+
+
+/**
+ * @title SnapshotToken
+ *
+ * @dev An ERC20 token which enables taking snapshots of accounts' balances.
+ * @dev This can be useful to safely implement voting weighed by balance.
+ */
+contract SnapshotToken is StandardToken {
+  using ArrayUtils for uint256[];
+
+  // The 0 id represents no snapshot was taken yet.
+  uint256 private currSnapshotId;
+
+  mapping (address => uint256[]) private snapshotIds;
+  mapping (address => uint256[]) private snapshotBalances;
+
+  event Snapshot(uint256 id);
+
+  function transfer(address _to, uint256 _value) public returns (bool) {
+    _updateSnapshot(msg.sender);
+    _updateSnapshot(_to);
+    return super.transfer(_to, _value);
+  }
+
+  function transferFrom(address _from, address _to, uint256 _value) public returns (bool) {
+    _updateSnapshot(_from);
+    _updateSnapshot(_to);
+    return super.transferFrom(_from, _to, _value);
+  }
+
+  function snapshot() public returns (uint256) {
+    currSnapshotId += 1;
```

Comment:
could use `AutoIncrementing` here if it happens to be available when this is merged